File: site/pipeline-status.js
// pipeline-status.js — live pipeline status card + manual "run now" trigger.
//
// Reads pipeline_status/current (written by the home machine via a service
// account) and renders what the sync run is doing right now, so new books can
// be checked on from anywhere instead of walking to the PC or waiting out the
// silent 8-hourly schedule.
//
// The trigger works by writing a request doc that a watcher on that machine
// polls for. Firestore rules make pipeline_requests create-only and UNREADABLE
// — see the note in firestore.rules. The token below is the only thing that
// makes a request real, so it is entered once by the admin and kept in
// localStorage; it is never baked into this file or committed.
//
// ES module, browser-native, no build step (matches the rest of site/).

import {
  collection, doc, addDoc, onSnapshot, query, orderBy, limit, getDocs,
} from 'https://www.gstatic.com/firebasejs/10.8.0/firebase-firestore.js';
import { col } from './fb-env.js';

const TOKEN_KEY = 'pipelineTriggerToken';

export function getToken() {
  try { return localStorage.getItem(TOKEN_KEY) || ''; } catch { return ''; }
}
export function setToken(v) {
  try { localStorage.setItem(TOKEN_KEY, (v || '').trim()); } catch { /* private mode */ }
}

const esc = (s) => String(s ?? '').replace(/[&<>"']/g,
  (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c]));

function ago(iso) {
  if (!iso) return '';
  const secs = (Date.now() - new Date(iso).getTime()) / 1000;
  if (Number.isNaN(secs)) return '';
  if (secs < 60) return `${Math.max(0, Math.round(secs))}s ago`;
  if (secs < 3600) return `${Math.round(secs / 60)}m ago`;
  if (secs < 86400) return `${Math.round(secs / 3600)}h ago`;
  return `${Math.round(secs / 86400)}d ago`;
}

function fmtDuration(sec) {
  if (!sec && sec !== 0) return '';
  if (sec < 60) return `${sec}s`;
  return `${Math.floor(sec / 60)}m ${sec % 60}s`;
}

const STEP_MARK = { done: '✓', active: '▸', failed: '✕', pending: '·', skipped: '–' };

/**
 * A run that never called finish_run (power cut, forced reboot) would otherwise
 * show "running" forever. Treat a stale heartbeat as unknown rather than live.
 */
function isStale(status) {
  if (status?.state !== 'running') return false;
  const t = new Date(status.updatedAt || 0).getTime();
  return Number.isFinite(t) && (Date.now() - t) > 15 * 60 * 1000;
}

function renderSteps(status) {
  return (status.steps || []).map((s) => {
    const isActive = s.state === 'active';
    const cls = `pl-step pl-step--${esc(s.state)}`;
    const detail = s.detail ? `<span class="pl-step__detail">${esc(s.detail)}</span>` : '';
    let bar = '';
    if (isActive && status.progress && status.stepKey === 'upload') {
      const p = status.progress;
      const pct = Math.max(0, Math.min(100, Number(p.pct) || 0));
      bar = `
        <div class="pl-progress">
          <div class="pl-progress__bar"><span style="width:${pct}%"></span></div>
          <div class="pl-progress__label">
            ${esc(p.file)} · ${pct}%${p.sizeMb ? ` · ${esc(p.sizeMb)} MB` : ''}
            ${p.total ? ` · file ${esc(p.index)}/${esc(p.total)}` : ''}
          </div>
        </div>`;
    }
    return `<li class="${cls}">
        <span class="pl-step__mark">${STEP_MARK[s.state] || '·'}</span>
        <span class="pl-step__label">${esc(s.label)}</span>
        ${detail}
      </li>${bar}`;
  }).join('');
}

function renderSummary(status) {
  const s = status.summary || {};
  const bits = [];
  if (s.idle) bits.push('nothing new to upload');
  if (s.uploaded) bits.push(`${s.uploaded} uploaded`);
  if (s.failed) bits.push(`${s.failed} failed`);
  if (s.books) bits.push(`${s.books} books total`);
  if (Array.isArray(s.newBooks) && s.newBooks.length) {
    bits.push(`new: ${s.newBooks.slice(0, 3).map(esc).join(', ')}${s.newBooks.length > 3 ? '…' : ''}`);
  }
  return bits.length ? `<div class="pl-summary">${bits.join(' · ')}</div>` : '';
}

export function renderStatus(el, status) {
  if (!status) {
    el.innerHTML = `<div class="pl-card pl-card--idle">
      <div class="pl-head"><span class="pl-dot pl-dot--idle"></span><strong>Pipeline</strong>
      <span class="pl-state">no runs recorded yet</span></div>
      <p class="pl-hint">The status card fills in the first time the pipeline runs
      with credentials configured.</p></div>`;
    return;
  }

  const stale = isStale(status);
  const state = stale ? 'unknown' : (status.state || 'unknown');
  const label = {
    running: 'RUNNING', success: 'SUCCESS', partial: 'PARTIAL',
    failed: 'FAILED', unknown: 'NO HEARTBEAT',
  }[state] || state.toUpperCase();

  const when = status.state === 'running' && !stale
    ? `started ${ago(status.startedAt)}`
    : `${ago(status.finishedAt || status.updatedAt)}${status.durationSec ? ` · took ${fmtDuration(status.durationSec)}` : ''}`;

  el.innerHTML = `
    <div class="pl-card pl-card--${esc(state)}">
      <div class="pl-head">
        <span class="pl-dot pl-dot--${esc(state)}"></span>
        <strong>Pipeline</strong>
        <span class="pl-state">${esc(label)}</span>
        <span class="pl-when">${esc(when)}</span>
        <span class="pl-trigger">${esc(status.trigger || '')}</span>
      </div>
      ${stale ? `<div class="pl-error">No heartbeat for over 15 minutes — the run may
        have been interrupted. Check output_files/pipeline_8h.log on the machine.</div>` : ''}
      ${status.error ? `<div class="pl-error">${esc(status.error)}</div>` : ''}
      <ul class="pl-steps">${renderSteps(status)}</ul>
      ${renderSummary(status)}
    </div>`;
}

export function watchStatus(db, el) {
  const ref = doc(collection(db, col('pipeline_status')), 'current');
  return onSnapshot(
    ref,
    (snap) => renderStatus(el, snap.exists() ? snap.data() : null),
    (err) => {
      el.innerHTML = `<div class="pl-card pl-card--failed">
        <div class="pl-head"><span class="pl-dot pl-dot--failed"></span><strong>Pipeline</strong>
        <span class="pl-state">STATUS UNAVAILABLE</span></div>
        <div class="pl-error">${esc(err.message || err)}</div></div>`;
    },
  );
}

export async function loadHistory(db, el, n = 8) {
  try {
    const q = query(collection(db, col('pipeline_runs')), orderBy('startedAt', 'desc'), limit(n));
    const snap = await getDocs(q);
    if (snap.empty) { el.innerHTML = ''; return; }
    const rows = snap.docs.map((d) => {
      const r = d.data();
      const s = r.summary || {};
      const outcome = s.idle ? 'nothing new'
        : [s.uploaded ? `${s.uploaded} uploaded` : '', s.books ? `${s.books} books` : '']
          .filter(Boolean).join(' · ') || '—';
      return `<tr>
        <td>${esc(new Date(r.startedAt).toLocaleString())}</td>
        <td class="pl-hist__state pl-hist__state--${esc(r.state)}">${esc(r.state)}</td>
        <td>${esc(r.trigger || '')}</td>
        <td>${esc(fmtDuration(r.durationSec))}</td>
        <td>${outcome}</td>
      </tr>`;
    }).join('');
    el.innerHTML = `<h3 class="pl-hist__title">Recent runs</h3>
      <table class="pl-hist">
        <thead><tr><th>Started</th><th>Result</th><th>Trigger</th><th>Took</th><th>Outcome</th></tr></thead>
        <tbody>${rows}</tbody>
      </table>`;
  } catch (e) {
    el.innerHTML = `<p class="pl-hint">Could not load history: ${esc(e.message || e)}</p>`;
  }
}

/**
 * Queue a manual run. Resolves to a human-readable confirmation.
 * Rejects if no token has been set — without it the watcher discards the request.
 */
export async function requestRun(db, requestedBy) {
  const token = getToken();
  if (!token || token.length < 16) {
    throw new Error('No trigger token saved. Paste the PIPELINE_TRIGGER_TOKEN from .env below first.');
  }
  await addDoc(collection(db, col('pipeline_requests')), {
    token,
    requestedAt: new Date().toISOString(),
    requestedBy: String(requestedBy || 'admin').slice(0, 80),
  });
  return 'Requested — the machine checks every 3 minutes, so it starts within ~3 min (or is skipped if a run is already going or the cooldown is active).';
}

File: app/pipeline_status.py
# app/pipeline_status.py
# Publish live pipeline status to Firestore so the admin panel can show what
# the 8-hourly run is doing without anyone sitting at the machine.
#
# Design rules:
#   1. NEVER raise. The pipeline moves audiobooks and deploys a site; a status
#      backend being down, misconfigured or absent must not cost a run. Every
#      public function swallows its own errors and degrades to a no-op.
#   2. No credentials => silent no-op. A fresh clone with no service account
#      behaves exactly as before this module existed.
#   3. Writes are throttled (see MIN_WRITE_INTERVAL_S) because step 4 reports
#      upload progress per chunk and Firestore bills per write.
#
# Collections (lane-suffixed to match site/fb-env.js — the /dev/ Pages lane and
# localhost read *_dev, so a dev run never overwrites the prod status card):
#
#   pipeline_status/current   one doc, overwritten — what is happening NOW
#   pipeline_runs/{run_id}    one doc per run — history for the panel
#
# Credentials: a Firebase service-account JSON at scripts/firebase_service_account.json
# (gitignored), or the path in FIREBASE_SERVICE_ACCOUNT. See docs/access/FIREBASE.md.

from __future__ import annotations

import os
import time
import traceback
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from app.config import PROJECT_ROOT

# Step 0-6 mirror scripts/sync_to_drive.py run_pipeline(). Kept here so the UI
# can render the whole list greyed out before the run reaches each one.
STEPS: list[tuple[str, str]] = [
    ("audit", "Purchase audit"),
    ("sort", "Sort books"),
    ("detect", "Detect new books"),
    ("folders", "Read Drive folders"),
    ("upload", "Upload to Drive"),
    ("catalog", "Rebuild catalog"),
    ("publish", "Commit & deploy"),
]

DEFAULT_KEY_PATH = PROJECT_ROOT / "scripts" / "firebase_service_account.json"
MIN_WRITE_INTERVAL_S = 3.0  # throttle upload-progress chatter

_db: Any = None
_init_tried = False
_run_id: str | None = None
_run_started: str | None = None
_state: dict[str, Any] = {}
_last_write = 0.0
_disabled_reason: str | None = None


def _lane_suffix() -> str:
    """'' for prod, '_dev' for dev runs — mirrors site/fb-env.js col()."""
    return "_dev" if os.getenv("PIPELINE_LANE", "").lower() == "dev" else ""


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _client():
    """Lazily build the Firestore client. Returns None if unavailable."""
    global _db, _init_tried, _disabled_reason
    if _init_tried:
        return _db
    _init_tried = True

    key_path = Path(os.getenv("FIREBASE_SERVICE_ACCOUNT") or DEFAULT_KEY_PATH)
    if not key_path.exists():
        _disabled_reason = f"no service account at {key_path}"
        return None

    try:
        import firebase_admin
        from firebase_admin import credentials, firestore

        if not firebase_admin._apps:
            firebase_admin.initialize_app(credentials.Certificate(str(key_path)))
        _db = firestore.client()
    except ImportError:
        _disabled_reason = "firebase-admin not installed (pip install firebase-admin)"
    except Exception as e:  # bad key, no network, project mismatch …
        _disabled_reason = f"{type(e).__name__}: {e}"
    return _db


def _push(force: bool = False) -> None:
    """Write _state to pipeline_status/current. Throttled unless force."""
    global _last_write
    db = _client()
    if db is None:
        return
    if not force and (time.time() - _last_write) < MIN_WRITE_INTERVAL_S:
        return
    try:
        db.collection(f"pipeline_status{_lane_suffix()}").document("current").set(_state)
        _last_write = time.time()
    except Exception:
        pass  # never let telemetry break a run


# --------------------------------------------------------------------------
# Public API — all no-ops when Firestore is unavailable
# --------------------------------------------------------------------------
def start_run(trigger: str = "scheduled") -> str:
    """Begin a run. trigger: 'scheduled' | 'manual' | 'cli'."""
    global _run_id, _run_started, _state
    try:
        _run_started = _now()
        _run_id = _run_started.replace(":", "").replace("-", "")[:15]
        _state = {
            "runId": _run_id,
            "state": "running",
            "trigger": trigger,
            "startedAt": _run_started,
            "updatedAt": _run_started,
            "host": os.getenv("COMPUTERNAME") or "unknown",
            "stepIndex": -1,
            "stepKey": None,
            "stepLabel": None,
            "steps": [{"key": k, "label": lbl, "state": "pending", "detail": ""} for k, lbl in STEPS],
            "progress": None,
            "error": None,
            "summary": {},
        }
        _push(force=True)
    except Exception:
        pass
    return _run_id or ""


def step(key: str, detail: str = "") -> None:
    """Mark a step active; everything before it becomes done."""
    try:
        if not _state:
            return
        idx = next((i for i, (k, _) in enumerate(STEPS) if k == key), None)
        if idx is None:
            return
        for i, s in enumerate(_state["steps"]):
            if i < idx and s["state"] in ("pending", "active"):
                s["state"] = "done"
            elif i == idx:
                s["state"] = "active"
                if detail:
                    s["detail"] = detail
        _state.update(
            stepIndex=idx, stepKey=key, stepLabel=STEPS[idx][1],
            progress=None, updatedAt=_now(),
        )
        _push(force=True)
    except Exception:
        pass


def step_detail(key: str, detail: str) -> None:
    """Update a step's one-line detail ('3 sorted', '397 folders cached')."""
    try:
        if not _state:
            return
        for s in _state["steps"]:
            if s["key"] == key:
                s["detail"] = detail
        _state["updatedAt"] = _now()
        _push(force=True)
    except Exception:
        pass


def upload_progress(file_name: str, pct: int, index: int, total: int, size_mb: float = 0.0) -> None:
    """Per-file upload progress. Throttled — safe to call every chunk."""
    try:
        if not _state:
            return
        _state["progress"] = {
            "file": file_name, "pct": int(pct), "index": index,
            "total": total, "sizeMb": round(size_mb, 1),
        }
        _state["updatedAt"] = _now()
        _push()  # throttled
    except Exception:
        pass


def set_summary(**fields: Any) -> None:
    """Accumulate run totals (uploaded, sorted, books, newBooks …)."""
    try:
        if not _state:
            return
        _state.setdefault("summary", {}).update(fields)
        _state["updatedAt"] = _now()
        _push(force=True)
    except Exception:
        pass


def finish_run(state: str = "success", error: str | None = None) -> None:
    """Close the run: flip the status doc and append to pipeline_runs history."""
    try:
        if not _state:
            return
        for s in _state["steps"]:
            if s["state"] == "active":
                s["state"] = "done" if state == "success" else "failed"
        # An idle run legitimately stops at step 2. Leaving the rest "pending"
        # on a SUCCESS card reads as though the run stalled, so distinguish
        # "never reached, and that's fine" from "still to come".
        if state != "running":
            for s in _state["steps"]:
                if s["state"] == "pending":
                    s["state"] = "skipped"
        _state.update(
            state=state, error=error, progress=None,
            finishedAt=_now(), updatedAt=_now(),
        )
        if _run_started:
            try:
                started = datetime.fromisoformat(_run_started)
                _state["durationSec"] = round(
                    (datetime.now(timezone.utc) - started).total_seconds()
                )
            except Exception:
                pass
        _push(force=True)

        db = _client()
        if db is not None and _run_id:
            try:
                db.collection(f"pipeline_runs{_lane_suffix()}").document(_run_id).set(_state)
            except Exception:
                pass
    except Exception:
        pass


def fail_run(exc: BaseException) -> None:
    """Record a crash, including the traceback tail, then close the run."""
    try:
        tail = "".join(traceback.format_exception_only(type(exc), exc)).strip()
        finish_run(state="failed", error=tail[:500])
    except Exception:
        pass


# --------------------------------------------------------------------------
# Fine-grained manual step controls (owner ask 2026-08-16, catalog-platform
# /status Operations section: "give us fine control over each part of the
# pipeline in case we need to do part way steps"). See
# scripts/sync_to_drive.py's run_step()/STEP_INFO for the dispatcher this
# backs.
# --------------------------------------------------------------------------
def start_step_run(step_key: str, step_label: str, trigger: str) -> str:
    """Begin a STANDALONE single-step run. Deliberately NOT start_run(): that
    function scaffolds the full 7-entry STEPS list and its companion step()
    marks every entry BEFORE the given index 'done' on the assumption of
    sequential progression through the whole pipeline — true for a full run,
    false for an isolated step. Running 'upload' alone must never make the
    status page claim 'sort' also ran just now. So this scaffolds only the
    ONE step that is actually running, already 'active' (no separate call to
    step() needed). finish_run() itself needs no change: it walks
    _state['steps'] generically by state, not by position, so closing a
    single-step card correctly marks that one step done/failed and leaves
    nothing 'pending' to fall through to 'skipped'."""
    global _run_id, _run_started, _state
    try:
        _run_started = _now()
        _run_id = _run_started.replace(":", "").replace("-", "")[:15]
        _state = {
            "runId": _run_id,
            "state": "running",
            "trigger": trigger,
            "startedAt": _run_started,
            "updatedAt": _run_started,
            "host": os.getenv("COMPUTERNAME") or "unknown",
            "stepIndex": 0,
            "stepKey": step_key,
            "stepLabel": step_label,
            "steps": [{"key": step_key, "label": step_label, "state": "active", "detail": ""}],
            "progress": None,
            "error": None,
            "summary": {},
        }
        _push(force=True)
    except Exception:
        pass
    return _run_id or ""


def force_upload_result(ok: bool, configured: bool, reachable: bool | None, message: str) -> None:
    """Publish the standalone 'force full upload to the shelf server' result
    (owner ask 2026-08-16; see scripts/sync_to_server.py) to ITS OWN doc,
    `shelf_upload_status/current` — deliberately NOT `pipeline_status/current`.
    That doc is the pipeline's own health signal (the /status page's primary
    'Automated Book Pipeline' row); overwriting it here would make a
    force-upload run masquerade as a pipeline outcome. Same never-raise,
    silent-no-op-without-credentials contract as every other function in this
    module — a status-backend outage must never cost the underlying transfer.
    """
    db = _client()
    if db is None:
        return
    try:
        state = (
            "not_configured" if not configured
            else "unreachable" if not reachable
            else "success" if ok
            else "failed"
        )
        doc = {
            "ok": ok,
            "state": state,
            "message": message,
            "updatedAt": _now(),
            "host": os.getenv("COMPUTERNAME") or "unknown",
        }
        db.collection(f"shelf_upload_status{_lane_suffix()}").document("current").set(doc)
    except Exception:
        pass


# --------------------------------------------------------------------------
# Single-flight lock reporting (2026-08-16, docs/info/ROLES.md §1c/§1d) —
# "a blocked run must FAIL LOUDLY AND VISIBLE, never silently no-op — print
# who holds the lock and since when, and write it to pipeline_status so the
# /status page can show it." These write pipeline_status/current directly
# (self-contained snapshots; they don't go through start_run()/_state since
# no run actually began) and, for the abandon case, a pipeline_runs history
# entry too — a skipped scheduled slot must be as visible in history as a
# failed run, never a silent gap. Same "never raise" contract as the rest of
# this module.
# --------------------------------------------------------------------------
def blocked_run(trigger: str, holder_desc: str) -> None:
    """A run refused to start outright because app/core/pipeline_lock is
    held by someone else. Used by every non-scheduled entry point (manual,
    --rebuild-only, the watcher's manual trigger) — none of them retry, they
    fail immediately. See deferral_abandoned() for the scheduled trigger,
    which retries for up to 2h before giving up."""
    global _state
    try:
        now = _now()
        _state = {
            "runId": None,
            "state": "blocked",
            "trigger": trigger,
            "startedAt": now,
            "updatedAt": now,
            "finishedAt": now,
            "host": os.getenv("COMPUTERNAME") or "unknown",
            "stepIndex": -1,
            "stepKey": None,
            "stepLabel": None,
            "steps": [],
            "progress": None,
            "error": f"Blocked: pipeline lock held by {holder_desc}. Refused to start (trigger={trigger} does not retry).",
            "summary": {},
        }
        _push(force=True)
    except Exception:
        pass


def deferring(trigger: str, holder_desc: str, deferring_since: str, attempt: int, note: str = "") -> None:
    """Heartbeat published while the SCHEDULED trigger retries against a
    held lock (app/core/pipeline_schedule.py). Overwrites
    pipeline_status/current on every attempt; no history entry yet — one is
    written only once the deferral resolves, either here-adjacent via
    deferral_abandoned() or by the normal start_run()/finish_run() pair once
    the lock clears and the real run begins."""
    global _state
    try:
        now = _now()
        _state = {
            "runId": None,
            "state": "deferred",
            "trigger": trigger,
            "startedAt": deferring_since,
            "updatedAt": now,
            "finishedAt": None,
            "host": os.getenv("COMPUTERNAME") or "unknown",
            "stepIndex": -1,
            "stepKey": None,
            "stepLabel": None,
            "steps": [],
            "progress": None,
            "error": f"Deferred (attempt {attempt}): blocked by {holder_desc}. {note}".strip(),
            "summary": {"deferringSince": deferring_since, "attempt": attempt},
        }
        _push(force=True)
    except Exception:
        pass


def deferral_abandoned(trigger: str, holder_desc: str, deferring_since: str, attempts: int, recovered: bool = False) -> None:
    """The 2h defer window elapsed still blocked (or — 'recovered' — a
    crashed deferral's overdue abandonment is being surfaced late by a
    later invocation, so it is never silently lost). Writes BOTH the live
    card and a real pipeline_runs history entry: a skipped scheduled slot
    must be as visible in history as a failed run."""
    global _state
    try:
        now = _now()
        run_id = deferring_since.replace(":", "").replace("-", "")[:15] + "-deferred"
        if recovered:
            prefix = "RECOVERED (an earlier deferral crashed without logging its own abandonment): "
        else:
            prefix = ""
        error = (
            f"{prefix}ABANDONED after {attempts} retr{'y' if attempts == 1 else 'ies'} "
            f"over up to 2h: still blocked by {holder_desc} when the defer window "
            f"elapsed (deferring since {deferring_since}). This scheduled slot was "
            "skipped — the NEXT 8h slot proceeds normally on its original schedule."
        )
        _state = {
            "runId": run_id,
            "state": "skipped",
            "trigger": trigger,
            "startedAt": deferring_since,
            "updatedAt": now,
            "finishedAt": now,
            "host": os.getenv("COMPUTERNAME") or "unknown",
            "stepIndex": -1,
            "stepKey": None,
            "stepLabel": None,
            "steps": [],
            "progress": None,
            "error": error,
            "summary": {"deferringSince": deferring_since, "attempts": attempts},
        }
        try:
            started = datetime.fromisoformat(deferring_since)
            _state["durationSec"] = round((datetime.now(timezone.utc) - started).total_seconds())
        except Exception:
            pass
        _push(force=True)

        db = _client()
        if db is not None:
            try:
                db.collection(f"pipeline_runs{_lane_suffix()}").document(run_id).set(_state)
            except Exception:
                pass
    except Exception:
        pass


def status_note() -> str:
    """One line for the console so a silent no-op is visible in the log."""
    if _client() is not None:
        return f"[status] publishing to pipeline_status{_lane_suffix()}/current"
    return f"[status] disabled ({_disabled_reason})"
